### ISSUE TITLE
Add preconditions

### ISSUE THREAD

**Issue by dan-zheng** [tensorflow/swift-apis#517]
Preconditions (for shape-checking, valid value ranges, etc) would be valuable for better source locations in error messages and to avoid uninformative TensorFlow runtime errors.

Example: `matmul(_:_:)` currently crashes with an uninformative stack trace.
```swift
import TensorFlow
let x = Tensor([1, 1])
print(matmul(x, x)) // invalid
```
```
$ swift matmul.swift
Fatal error: In[0] is not a matrix. Instead it has shape [2]: file /usr/local/src/swift-build/tensorflow-swift-apis/Sources/TensorFlow/Bindings/EagerExecution.swift, line 300 # unuseful source location
Stack dump: ...
```

**Comment by Shashi456** [comment]
Could we start with making a list of what methods need a precondition ?

**Comment by dan-zheng** [comment]
> Could we start with making a list of what methods need a precondition ?

I'd recommend looking through files in `Sources/TensorFlow/Operators` and `Sources/TensorFlow/Layers` and look for "precondition" comments on properties/parameters that aren't actually checked.

Examples:
* `func conv2D`, [``/// - Precondition: `input` must have rank `4`.``](https://github.com/tensorflow/swift-apis/blob/27ea161a69ed207c26028286ce751888c6f9a12d/Sources/TensorFlow/Operators/NN.swift#L83)
  * Missing a precondition at top of function.
* `Conv1D.filter`, [`/// The 3-D convolution filter.`](https://github.com/tensorflow/swift-apis/blob/27ea161a69ed207c26028286ce751888c6f9a12d/Sources/TensorFlow/Layers/Convolutional.swift#L21)
  * Missing precondition in initializer.
* Other shape-related functions: `matmul`, `reshape`, etc
* Probability-related APIs (`Tensor.droppingOut`, `Dropout` layer) should have `0..<1` preconditions for probability values.

**Comment by Shashi456** [comment]
@dan-zheng I'll try to put up a list in the next few hours

**Comment by ocampor** [comment]
@dan-zheng I can create a runtime performance analysis on some of the functions with preconditions. Do you have any preferred methodology to assess the impact? Any recommendation to make the analysis more reliable?

**Comment by Shashi456** [comment]
A list of preconditions :
- [ ] 
- [ ] 
- [ ] 
- [ ]

**Comment by ocampor** [comment]
Today I’m going to spend some time looking the missing preconditions.

**Comment by rickwierenga** [comment]
@Shashi456, is your to do list (https://github.com/tensorflow/swift-apis/issues/517#issuecomment-537128724) up to date?

**Comment by SumanSudhir** [comment]
> @Shashi456, is your to do list ([#517 (comment)](https://github.com/tensorflow/swift-apis/issues/517#issuecomment-537128724)) up to date?

No, it is not updated, preconditions have been added to almost every function mentioned in the list https://github.com/tensorflow/swift-apis/issues/517#issuecomment-537128724

**Comment by t-ae** [comment]
There's no `Layer`s in list. Can you add them @Shashi456?

---

Unlike functions, there are multiple choice in `Layer`s where we check.

1. During `init`
2. When parameters are assigned (in `didSet`)
3. During `callAsFunction`

The checks depend on input must go into `callAsFunction`.
The other checks like `filter.rank == 4` of `Conv2D` can be done outside `callAsFunction`.
I think it's good to have them in `init`.

But, since many parameters are `public`, users can assign wrong tensor after `init`.
Should we consider that?

If we consider, we have to have same checks done in `init` in `callAsFunction` redundantly.

We can have them in `didSet` and I think it is besically better, but it keeps `oldValue` for now, so memory consumption will be double.
We should wait [SE-0268](https://github.com/apple/swift-evolution/blob/master/proposals/0268-didset-semantics.md) to be merged.

**Comment by Shashi456** [comment]
@t-ae I'll try to add some more functions for this.
